File: array/SlidingWindow/variable_size/java/SlidingWindow.java
public final class SlidingWindow {
    private SlidingWindow() {}

    private static boolean addOverflow(long a, long b) {
        return (b > 0 && a > Long.MAX_VALUE - b) || (b < 0 && a < Long.MIN_VALUE - b);
    }

    private static boolean subOverflow(long a, long b) {
        return (b < 0 && a > Long.MAX_VALUE + b) || (b > 0 && a < Long.MIN_VALUE + b);
    }

    public static Result minWindowAtLeastTarget(long[] values, long target) {
        long windowSum = 0;
        int left = 0;
        int bestLength = -1;
        int bestLeft = -1;
        int bestRight = -1;

        for (int right = 0; right < values.length; right++) {
            if (addOverflow(windowSum, values[right])) {
                return null;
            }
            windowSum += values[right];

            while (windowSum >= target) {
                int currentLength = right - left + 1;
                if (bestLength == -1 || currentLength < bestLength) {
                    bestLength = currentLength;
                    bestLeft = left;
                    bestRight = right;
                }

                if (subOverflow(windowSum, values[left])) {
                    return null;
                }
                windowSum -= values[left];
                left++;
            }
        }

        return new Result(bestLength, bestLeft, bestRight);
    }

    public record Result(int bestLength, int bestLeft, int bestRight) {}
}

File: array/SlidingWindow/variable_size/python/SlidingWindow.py
MIN_I64 = -(1 << 63)
MAX_I64 = (1 << 63) - 1


def _checked_i64(value: int) -> int | None:
    if value < MIN_I64 or value > MAX_I64:
        return None
    return value


def min_window_at_least_target(values: list[int], target: int) -> tuple[int, int, int] | None:
    window_sum = 0
    left = 0
    best_length = -1
    best_left = -1
    best_right = -1

    for right, value in enumerate(values):
        window_sum += value
        if _checked_i64(window_sum) is None:
            return None

        while window_sum >= target:
            current_length = right - left + 1
            if best_length == -1 or current_length < best_length:
                best_length = current_length
                best_left = left
                best_right = right
            window_sum -= values[left]
            if _checked_i64(window_sum) is None:
                return None
            left += 1

    return best_length, best_left, best_right
